### ISSUE TITLE
Add versions of Add/Remove/Update ranges that takes an IAsyncEnumerable

### ISSUE THREAD

**Issue by EricStG** [dotnet/efcore#21837]
With EF 5 targeting netstandard 2.1, I'd like to be able to pass an `IAsyncEnumerable<T>` to AddRange, RemoveRange and UpdateRange so I don't have to force `ToListAsync()` on the enumerable before passing it to the method.

I would like to see something like the following additions to DbSet and DbContext:
```cs
public virtual System.Threading.Tasks.Task AddRangeAsync(IAsyncEnumerable<TEntity> entities, System.Threading.CancellationToken cancellationToken = null)
public virtual System.Threading.Tasks.Task RemoveRangeAsync(IAsyncEnumerable<TEntity> entities, System.Threading.CancellationToken cancellationToken = null)
public virtual System.Threading.Tasks.Task UpdateRangeAsync(IAsyncEnumerable<TEntity> entities, System.Threading.CancellationToken cancellationToken = null)
```

**Comment by roji** [comment]
@EricStG if I'm correctly understanding what you're asking for, these method would internally simply do:

```c#
await foreach (var i in entities) {
    ctx.Add(i);
}
```

The important thing to note here, is that these operations in EF Core don't perform any sort of async I/O, so the only reason to have these new async methods would be as very thin sugar which iterates through the IAsyncEnumerable you give them. I personally find this to be of little value, and it would increase the API surface which isn't a good thing.

To put it another way, I don't believe these methods belong on DbSet or DbContext any more than they belong on .NET's List, or indeed any other type which has an Add API.

**Comment by EricStG** [comment]
That's fair. I thought `AddRange` had some internal optimization over `Add`, but it's probably just sugar either way.

**Comment by roji** [comment]
I don't believe that's the case, but @ajcvickers may have additional info here.

**Comment by ajcvickers** [comment]
@EricStG In EF6, calling Add multiple times would also result in multiple calls to DetectChanges. AddRange avoided this. 

In EF Core, neither Add nor AddRange call DetectChanges, so there is little advantage to using AddRange.

**Comment by EricStG** [comment]
Good to know, closing then!
